driver-core/src/main/com/mongodb/internal/connection/LoggingCommandEventSender.java

```diff
@@ -71,12 +72,33 @@ class LoggingCommandEventSender implements CommandEventSender {
     @Override
     public void sendStartedEvent() {
         if (loggingRequired()) {
-            String commandString = redactionRequired ? format("{\"%s\": ...", commandName) : getTruncatedJsonCommand();
-            // TODO: log RequestContext?
-            logger.debug(
-                    format("Sending command '%s' with request id %d to database %s on connection [%s] to server %s",
-                            commandString, message.getId(),
-                            message.getNamespace().getDatabaseName(), description.getConnectionId(), description.getServerAddress()));
+            String commandString = redactionRequired ? "{}" : getTruncatedJsonCommand(commandDocument);
+            if (description.getServiceId() == null) {
+                logger.debug("Command started", getClusterId(),
+                        "Command \"%s\" started on database %s using a connection with driver-generated ID %d and server-generated ID %d "
+                                + "to %s:%s. The request ID is %s. Command: %s",
+                        "commandName", commandName,
+                        "databaseName", message.getNamespace().getDatabaseName(),
+                        "driverConnectionId", description.getConnectionId().getLocalValue(),
+                        "serverConnectionId", description.getConnectionId().getServerValue(),
+                        "serverHost", description.getServerAddress().getHost(),
+                        "serverPort", description.getServerAddress().getPort(),
+                        "requestId", message.getId(),
+                        "command", commandString);
+            } else {
+                logger.debug("Command started", getClusterId(),
+                        "Command \"%s\" started on database %s using a connection with driver-generated ID %d and server-generated ID %d "
+                                + "to %s:%s with service ID %s. The request ID is %s. Command: %s",
+                        "commandName", commandName,
+                        "databaseName", message.getNamespace().getDatabaseName(),
+                        "driverConnectionId", description.getConnectionId().getLocalValue(),
+                        "serverConnectionId", description.getConnectionId().getServerValue(),
+                        "serverHost", description.getServerAddress().getHost(),
+                        "serverPort", description.getServerAddress().getPort(),
+                        "serviceId", description.getServiceId(),
+                        "requestId", message.getId(),
+                        "command", commandString);
+            }
```

Comment:
1. This unfortunate significant code duplication is caused by the [spec requirement](https://github.com/mongodb/specifications/blob/master/source/logging/logging.rst#omitting-null-values-from-log-messages). If we had two logged values that may independently be `null`, would we quadruplicate the logging code?
2. The question is especially relevant because we log `description.getConnectionId().getServerValue()`, and `ConnectionId.getServerValue` is documented as `@Nullable`. Do we know that in all situations when we currently call `StructuredLogger.debug`, the server value is not `null`? If yes, could we add an assertion for it?

Reply:
1.  Yes, I think we would
2. I'd be uncomfortable adding this assertion, as I think it would be the first of its kind for this value and may expose differences in implementations of various server implementation (mongocryptd, Data Lake, etc).  If the value is actually null, I don't think it actually causes a problem in production, but only if there was an assertion in the spec tests (which there isn't)

Reply:
1. Do we want to somehow rise a concern about the spec requirement based on such consequences?
1.1. [regardless of what the answer to the previous question is] Do you actually think this is fine?
2. If we are not adding such an assertion because it may fail, then shouldn't we add two more logging branches to cover the nullability of `serverConnectionId` in addition to the nullability of `serviceId`?

Reply:
I think we're out of this mess at the cost of some extra allocations of Entry instances.  PTAL